presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -248,11 +253,20 @@ public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
         masterBuffer.addPages(serializedPageReferences);
 
         // process any pending reads from the client buffers
-        for (ClientBuffer clientBuffer : safeGetBuffersSnapshot()) {
+        List<ClientBuffer> buffers = safeGetBuffersSnapshot();
+        if (buffers.isEmpty()) {
+            return;
+        }
+        // handle potential for racy update of next index and client buffers present
+        int index = nextClientBufferIndex.get() % buffers.size();
+        for (int i = 0; i < buffers.size(); i++) {
             if (masterBuffer.isEmpty()) {
+                // Resume from the current client buffer on the next iteration
+                nextClientBufferIndex.set(index);
                 break;
             }
-            clientBuffer.loadPagesIfNecessary(masterBuffer);
+            buffers.get(index).loadPagesIfNecessary(masterBuffer);
```

Comment:
It looks to me the only difference with https://github.com/prestosql/presto/pull/2788 is the position of these two lines (line 268 - line 269). In https://github.com/prestosql/presto/pull/2788/, these two lines are at the beginning of for-loop (before the `if (masterBuffer.isEmpty())`) statement). 

Curious what's the discussion over this? :)

Reply:
Update: I think the discussion is here: https://github.com/prestodb/presto/pull/14083#discussion_r378451149